### PR TITLE
ddl: worker add ctxPool field to make executing sql in ddl package possible.

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve? 

Before this PR, we can not new a session to execute SQL in ddl package, it will lead to a cycle import because of session package import domain package, and domain package import ddl package. We need to execute SQL in ddl package in some cases. For example, we want to store history ddl job into a system table.

After this pr, we can execute a SQL as below:

```
        sessCtx, err := w.getSessionCtx()
	if err != nil {
		return ver, errors.Trace(err)
	}
	defer w.putSessionCtx(sessCtx)
	s := sessCtx.(sqlexec.SQLExecutor)
	_, err = s.Execute(context.Background(), "create table t(a int);")
	if err != nil {
		return ver, errors.Trace(err)
	}
```

### What is changed and how it works?

Use ctxPool to new a session, rather than import session package.

### Check List 

Tests 

 - No code

Code changes

 - Has interface methods change

Side effects

 - Increased code complexity

### CODE REVIEW

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -108,6 +112,25 @@ func (w *worker) close() {
 	log.Infof("[ddl-%s] close DDL worker", w)
 }
 
+// getSessionCtx get sessionctx from context resource pool.
+// Please remember to call putSessionCtx after you getting the sessionctx.
+func (w *worker) getSessionCtx() (sessionctx.Context, error) {
+	resource, err := w.ctxPool.Get()
```
Comment (crazycs520) [inline ddl/ddl_worker.go]: check `if ctxPool == nil` first?

Comment (winkyao) [reply ddl/ddl_worker.go]: the ctxPool must be not nil.

Comment (crazycs520) [reply ddl/ddl_worker.go]: Why line#82 have `if ctxPool != nil` ?

Comment (winkyao) [reply ddl/ddl_worker.go]: Just test code will pass a nil ctxPool.

Comment (winkyao) [reply ddl/ddl_worker.go]: I'll return mockContext in this case.

**Comment by winkyao** [comment]
@crazycs520 PTAL

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -108,6 +113,30 @@ func (w *worker) close() {
...
+	ctx := resource.(sessionctx.Context)
+	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
+	ctx.GetSessionVars().InRestrictedSQL = true
+	return ctx, nil
+}
+
+// putSessionCtx return sessionctx to context resource pool.
+func (w *worker) putSessionCtx(ctx sessionctx.Context) {
+	if w.ctxPool != nil {
+		w.ctxPool.Put(ctx.(pools.Resource))
```
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: `w.ctxPool.Put(ctx)` works the same?

Comment (lamxTyler) [reply ddl/ddl_worker.go]: Should we reset the `Autocommit` and `InRestrictedSQL`?

Comment (winkyao) [reply ddl/ddl_worker.go]: @lamxTyler No needs, the session is only using to run  restricted-sql. Actually, in delRangeManager, we don't reset these fields before we put it into ctxPool.

Comment (winkyao) [reply ddl/ddl_worker.go]: @tiancaiamao sessionctx.Context don't implement Close method, but *session does.

**Comment by winkyao** [comment]
@zimulala I afraid that is not a good idea, delete_range just need to focus on GC, and new a sessionContext isn't its work.

File: ddl/ddl_worker.go (hunk 2 above)
Comment (tiancaiamao) [reply ddl/ddl_worker.go]: Oh, I see.

**Comment by winkyao** [comment]
@zimulala @tiancaiamao  PTAL

File: domain/domain.go [hunk 3]
```diff
@@ -503,7 +503,7 @@ func (do *Domain) Init(ddlLease time.Duration, sysFactory func(*Domain) (pools.R
 	sysFac := func() (pools.Resource, error) {
 		return sysFactory(do)
 	}
-	sysCtxPool := pools.NewResourcePool(sysFac, 2, 2, resourceIdleTimeout)
+	sysCtxPool := pools.NewResourcePool(sysFac, 4, 10, resourceIdleTimeout)
```
Comment (shenli) [inline domain/domain.go]: Why change it?

Comment (winkyao) [reply domain/domain.go]: I misunderstand here, actually the maxCap of sysCtxPool need to be equal to the count of the ddl worker. So 2 is ok. I'll change it back.

**Comment by winkyao** [comment]
@shenli PTAL

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/session_generator.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/pingcap/tidb/util/mock"
+)
+
+// sessionGenerator is used to new session.
+type sessionGenerator struct {
+	resPool *pools.ResourcePool
+}
+
+// getSessionCtx gets sessionctx from context resource pool.
+// Please remember to call putSessionCtx after you getting the sessionctx.
```
Comment (coocood) [inline ddl/session_generator.go]: Do you mean `after you finished using sessionctx`?

File: ddl/ddl_worker.go [hunk 5]
```diff
@@ -104,6 +107,7 @@ func (w *worker) String() string {
 func (w *worker) close() {
 	close(w.quitCh)
 	w.delRangeManager.clear()
+	w.sessGenerator.close()
```
Comment (coocood) [inline ddl/ddl_worker.go]: Is it safe to close the session generator multiple times?

Comment (coocood) [reply ddl/ddl_worker.go]: And concurrently?

**Comment by winkyao** [comment]
@coocood PTAL

File: util/mock/context.go [hunk 6]
```diff
@@ -46,6 +49,11 @@ type Context struct {
 	pcache      *kvcache.SimpleLRUCache
 }
 
+// Execute implements sqlexec.SQLExecutor Execute interface.
+func (c *Context) Execute(ctx context.Context, sql string) ([]ast.RecordSet, error) {
+	return nil, nil
```
Comment (coocood) [inline util/mock/context.go]: It's better to return an error for the not supported method.

**Comment by winkyao** [comment]
@coocood PTAL

File: ddl/ddl_worker_test.go [hunk 7]
```diff
@@ -710,3 +711,23 @@ func (s *testDDLSuite) TestParallelDDL(c *C) {
...
+	worker := d.generalWorker()
+	c.Assert(worker, NotNil)
+
+	// In test environment, worker.ctxPool will be nil, and get will return mock.Context.
+	// We just test that can use it to call sqlexec.SQLExecutor.Execute.
+	sess, err := worker.sessPool.get()
+	c.Assert(err, IsNil)
+	defer worker.sessPool.put(sess)
+	se := sess.(sqlexec.SQLExecutor)
+	_, _ = se.Execute(context.Background(), "create table t(a int);")
```
Comment (zimulala) [inline ddl/ddl_worker_test.go]: Do we need to check the error？

Comment (winkyao) [reply ddl/ddl_worker_test.go]: No need, please see the comments in line 726

File: ddl/ddl_worker.go [hunk 8]
```diff
@@ -62,21 +62,24 @@ type worker struct {
 	quitCh   chan struct{}
 	wg       sync.WaitGroup
 
-	reorgCtx        *reorgCtx // reorgCtx is used for reorganization.
+	sessPool        *sessionPool // sessPool is used to new sessions to execute SQL in ddl package.
```
Comment (zhexuany) [inline ddl/ddl_worker.go]: this change seems not to be related to this PR.

Comment (winkyao) [reply ddl/ddl_worker.go]: why not?

**Comment by winkyao** [comment]
@coocood @zimulala PTAL

File: ddl/ddl_worker_test.go [hunk 9]
```diff
@@ -729,6 +729,5 @@ func (s *testDDLSuite) TestDDLPackageExecuteSQL(c *C) {
 	c.Assert(err, IsNil)
 	defer worker.sessPool.put(sess)
 	se := sess.(sqlexec.SQLExecutor)
-	_, err = se.Execute(context.Background(), "create table t(a int);")
-	c.Assert(err, IsNil)
+	_, _ = se.Execute(context.Background(), "create table t(a int);")
```
Comment (coocood) [inline ddl/ddl_worker_test.go]: The test should not use mock context.
`testNewDDL` passes nil `ctxPool`, we should use `newDDL` instead.

File: ddl/ddl_worker.go (hunk 8 above)
Comment (ciscoxll) [reply ddl/ddl_worker.go]: @zhexuany Here often need to use ` sessCtx `.

**Pull request merged**